<font size="+3" center><strong> Real Estate Pipeline — Step 3: Data Preprocessing</strong></font>



This notebook documents the full preprocessing pipeline for the Montreal real estate dataset.
All transformations are performed directly in PostgreSQL to leverage the database engine's 
performance and to keep the data layer clean and reproducible.

## Dataset Overview
- Source: ******* (Montreal Island)
- Scraping runs: multiple passes to approach listing saturation
- Tables: listings, properties, locations, brokers, price_history, extension tables, scrape_runs
- Views: v_listings, v_condo, v_plex, v_commercial

## Preprocessing Steps
0. JSON - extract field from jsonb columns
1. Base views — normalize joins across tables
2. Data quality audit — missing values, outliers, type issues
3. Cleaning — filters, null strategy, outlier caps
4. Feature engineering — derived columns ready for ML
5. Export — clean datasets per property type

In [1]:
# setup environment
import warnings
from dotenv import load_dotenv, find_dotenv
import os 

_ = load_dotenv(find_dotenv()) # load environment variables from .env file


PASSWORD = os.getenv('DB_PASSWORD') # get the database password from environment variables

%load_ext sql 
%config SqlMagic.style = '_DEPRECATED_DEFAULT' # set the style for SQL output
%config SqlMagic.autopandas = True
%config SqlMagic.displaylimit = 50
%config SqlMagic.feedback = False

connecyion_string = f"postgresql://postgres:{PASSWORD}@localhost/real_estate"

# Connect to the PostgreSQL database using credentials from environment variables
%sql {connecyion_string}





warnings.simplefilter(action="ignore", category=FutureWarning)

In [2]:
# Data manipulation libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns



pd.set_option("display.max_rows", None)  # Control rows displayed in DataFrame outputs
pd.set_option("display.max_columns", None)  # Show all columns in DataFrame outputs
# Plot defaults
sns.set_style("whitegrid")
plt.rcParams["figure.figsize"] = (12, 6)

## 0. JSON Field Extraction

Since the characteristics and financial data on the website are property-specific, the database design cannot include all possible fields due to their unpredictability. Those variable characteristics and financial data were recorded on JSON columns in the `properties` table. Therefore, we will first display all the fields that have been loaded in both columns.

In [14]:
%%sql
SELECT key, count(*) AS appearance_count, round((count(*) * 100.0 / (SELECT count(*) FROM properties)), 2) AS percentage
FROM properties, jsonb_object_keys(characteristics) AS key
GROUP BY key
ORDER BY appearance_count DESC; -- Overall distribution of characteristics keys [Full database]


 * postgresql://postgres:***@localhost/real_estate


,key,appearance_count,percentage
0,year_built,13002,85.30
1,move_in_date,12897,84.61
2,additional_features,9868,64.74
3,parking_total,7525,49.37
4,condominium_type,7146,46.88
5,net_area,6012,39.44
6,lot_area,4248,27.87
7,building_style,3870,25.39
8,use_of_property,3575,23.45
9,pool,3337,21.89


The json object tracks 26 characteristics features accross all listings. Since this validation is performed on the full dataset, missing values are largely attributable to property-type-specific differences rather than systematic data absence. However, some features require double additional verification. 
- `year_built`: This field requires further validation to determine whether missing values result from scraping issues or from the platform itself not providing the information.
- `property area` : This is a property-specific measure. For condos, it corresponds to net area; for houses, to lot area; and for plexes, to both living and lot areas. In the listing view design, a unified area column will be created following this hierarchy: net area for condos, living area for houses, and building area for plexes.
- ```parking_total```: The dataset distinguishes between driveway and garage parking spaces. These will be aggregated into a single `````parking_total````` column. Additionally, the garage indicator will be recast as a boolean variable to capture the presence or absence of a garage.
- `move_in_date`: It is useful to store this field in a dedicated column, as it may provide valuable insights during exploratory data analysis (EDA). 
- Others: All remaining characteristics are either already represented in dedicated columns or are sparsely populated (< 50%). The latter will be backfilled where appropriate. 

### Area Choice per Property Type (Pricing Model Input)


| Property Type | Primary Feature (Building) | Secondary Feature (Land) | Rationale |
| :--- | :--- | :--- | :--- |
| **Condo** | **Net Area** | *None / Common Element* | High accuracy; reflects real usable living space. |
| **House** | **Living Area** (Above-grade) | **Lot Area** | Price is driven by house size + yard value. |
| **Plex** | **Total Building Area** | **Lot Area** | Includes all rental units; land has development potential. |

Next: We shall check how the `financial_data` json column behaves. 

In [15]:
%%sql 
SELECT keys, count(*) AS appearance_count, round((count(*) * 100.0 / (SELECT count(*) FROM properties)), 2) AS percentage
FROM properties, jsonb_object_keys(financial_data) AS keys
GROUP BY keys
ORDER BY count(*) DESC; -- Overall distribution of financial_data keys [Full database]



 * postgresql://postgres:***@localhost/real_estate


,keys,appearance_count,percentage
0,taxes__total,10377,68.08
1,taxes__school_2025,7094,46.54
2,fees__total,6081,39.89
3,taxes__municipal_2026,6021,39.50
4,fees__condominium_fees,5858,38.43
5,municipal_assessment_(2026)__total,4000,26.24
6,municipal_assessment_(2026)__lot,3983,26.13
7,municipal_assessment_(2026)__building,3974,26.07
8,taxes__municipal_2025,3558,23.34
9,municipal_assessment_(2024)__total,2398,15.73


The JSON object contains 426 financial-related fields across all listings. However, many of these fields are duplicated due to differences in naming conventions, assessment years, and property types. This high variability will require harmonization, as well as backfilling for missing or previously unanticipated columns.

- ```municipal_assessment_(year)__extend```: This field encodes three levels of information: the type of assessment, the assessment year, and the assessment value. It will be necessary to parse and separate each component before extracting the relevant value.
- **Taxes and assessment data**: These are standard components in real estate listings and will require systematic backfilling to ensure completeness.

## 1. Base Views

We start by creating reusable views that normalize the joins between 
`listings`, `properties`, `locations` and extension tables. These views 
are the foundation for all subsequent analysis.

### 1.1 Residential View
This view filters residential for-sale listings and flattens the schema 
into a some ready-queryable structure. 

[__Personal Note__]
I struggled somewhat to choose between two strategies: backfilling missing data directly or modifying the database design by creating views that dynamically consolidate and standardize the relevant information. The second approach is more appealing to me, as it avoids repeating backfilling operations after each scraping run and provides a more maintainable and scalable solution.

In [ ]:
%%sql
CREATE OR REPLACE VIEW v_residential_sale AS
SELECT
    l.id                                as listing_id,
    l.listing_id                        as platform_id,
    l.price,                            
    l.created_at,
    l.updated_at,
    l.description,
    l.category,
    COALESCE(
        p.year_built, 
        nullif(regexp_replace(p.characteristics ->> 'year_built', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER,
        (CASE 
            WHEN l.description ~* 'new|construction' 
            THEN substring(l.description from '\y20[0-9]{2}\y')::INTEGER 
            ELSE NULL 
         END)
    ) AS year_built,
    p.property_type,
    COALESCE(
        p.lot_size_sqft, 
        NULLIF(regexp_replace(p.characteristics ->> 'lot_area', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER
    ) AS lot_size_sqft,
    COALESCE(
        NULLIF(regexp_replace(p.characteristics ->> 'net_area', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER,
        NULLIF(regexp_replace(p.characteristics ->> 'living_area', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER,
        NULLIF(regexp_replace(p.characteristics ->> 'gross_area', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER,
        NULLIF(regexp_replace(p.characteristics ->> 'building_area_at_ground_level', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER
    ) AS area_sqft,
    COALESCE(p.bedrooms, NULLIF(regexp_replace(p.characteristics ->> 'bedrooms', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER) AS bedrooms,
    COALESCE(p.bathrooms, NULLIF(regexp_replace(p.characteristics ->> 'bathrooms', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER) AS bathrooms,
    COALESCE((
        SELECT SUM(val[1]::int)
        FROM regexp_matches(p.characteristics ->> 'parking_total', '\(([0-9]+)\)', 'g') AS val
    ),p.parking, 0) AS parking_total,
    COALESCE((p.characteristics ->> 'parking_total') ~ 'Garage', p.garage, false) AS garage,
    COALESCE(NULLIF(regexp_replace(p.characteristics ->> 'move_in_date', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER, 0) AS move_in_date,
    COALESCE(NULLIF(regexp_replace(p.characteristics ->> 'total_rooms', '[^0-9]', '', 'g'), ''::TEXT)::INTEGER, p.total_rooms) AS total_rooms,
    p.pool,
    p.basement,
    COALESCE(
        substring(
            (
                SELECT key
                FROM jsonb_object_keys(p.financial_data) AS key
                WHERE key LIKE 'municipal_assessment%'
                ORDER BY key DESC
                LIMIT 1
            ) FROM '([0-9]+)'
    )::INTEGER,  EXTRACT( YEAR FROM l.created_at)::INTEGER
) AS assessment_year,
    p.building_assessment,
    p.lot_assessment,
    (
        SELECT value
        FROM jsonb_each_text(p.financial_data)
        WHERE key LIKe 'taxes__total%'
        ORDER BY key DESC
        LIMIT 1
    ) AS total_taxes,
   (
        SELECT value
        FROM jsonb_each_text(p.financial_data)
        WHERE key LIKe 'taxes__school%'
        ORDER BY key DESC
        LIMIT 1
    ) AS school_taxes,
    (
        SELECT value
        FROM jsonb_each_text(p.financial_data)
        WHERE key LIKe 'taxes__municipal%'
        ORDER BY key DESC
        LIMIT 1
    ) AS municipal_taxes,
    loc.civic_number,
    loc.street_name,
    loc.city,
    loc.borough,
    loc.latitude,
    loc.longitude
FROM listings AS l
JOIN properties AS p ON l.property_id = p.id
JOIN locations AS loc  ON p.location_id  = loc.id
WHERE l.category     = 'for_sale'
AND   p.property_type IN (
    'house', 'condo', 'duplex', 'triplex',
    'quadruplex', 'quintuplex', 'condominium_house',
    'loft_studio', 'cottage', 'mobile_home'
) ;

 * postgresql://postgres:***@localhost/real_estate


""


In [9]:
%%sql
CREATE OR REPLACE VIEW v_condo AS
SELECT
    vl.*,
    loc.unit_number,
    lc.locker,
    COALESCE(NULLIF(p.characteristics ->> 'condominium_type', ''::TEXT)) AS condo_type,
    COALESCE(
        substring(
            (
                SELECT value
                FROM jsonb_each_text(p.financial_data)
                WHERE key LIKE 'fees__condo%'
                ORDER BY key DESC
                LIMIT 1
            ) FROM '([0-9]+)'
    )::NUMERIC, null) AS condo_fees_yearly
FROM v_residential_sale AS vl
JOIN listings l ON vl.listing_id = l.id
JOIN properties p ON l.property_id = p.id
JOIN listing_condo AS lc ON vl.listing_id = lc.listing_id
JOIN locations AS loc ON p.location_id = loc.id
WHERE vl.property_type IN ('condo', 'loft_studio', 'condominium_house');


 * postgresql://postgres:***@localhost/real_estate


""


In [10]:
%%sql  
CREATE OR REPLACE VIEW v_plex AS
SELECT
    vl.*,
    lp.units,
    lp.rental_income,
    COALESCE(NULLIF(regexp_replace(p.characteristics ->> 'living_area', '[^0-9.]', '', 'g'), ''), null)::NUMERIC AS living_area,
    COALESCE(NULLIF(regexp_replace(p.characteristics ->> 'lot_area', '[^0-9.]', '', 'g'), ''), null)::NUMERIC AS lot_area,
    COALESCE(
        NULLIF(p.characteristics ->> 'residential_units', ''::TEXT)::TEXT,
        NULLIF(p.characteristics ->> 'residential_unit', ''::TEXT)::TEXT
        ) AS residential_units
FROM v_residential_sale AS vl
JOIN listings l ON vl.listing_id = l.id
JOIN properties p ON l.property_id = p.id
JOIN listing_plex AS lp ON l.id = lp.listing_id
WHERE vl.property_type IN ('duplex', 'triplex', 'quadruplex', 'quintuplex', 'multiplex');

 * postgresql://postgres:***@localhost/real_estate


""


In [ ]:
%%sql
CREATE OR REPLACE VIEW v_commercial AS
SELECT
    vl.listing_id,
    vl.price,
    vl.category, -- 'for_sale' or 'for_rent' (Lease)
    -- 1. Space Metrics
    COALESCE(
        (p.characteristics ->> 'available_area')::numeric,
        core.area_sqft
    ) AS usable_sqft,
    p.characteristics ->> 'zoning_type' AS zoning, -- e.g., Industrial, Retail, Office
    
    -- 2. Logistics
    (p.characteristics ->> 'clear_height')::numeric AS ceiling_height_ft,
    (p.characteristics ->> 'drive_in_doors')::int AS loading_doors,
    (p.characteristics ->> 'has_rail_access')::boolean AS rail_access,
    
    -- 3. Financial / Business
    p.financial_data ->> 'lease_type' AS lease_type, -- e.g., Triple Net (NNN), Gross
    COALESCE(NULLIF(regexp_replace(p.financial_data ->> 'additional_rent', '[^0-9.]', '', 'g'), ''), '0')::numeric AS add_rent_sqft,
    
    -- 4. Common Commercial Fields
    p.characteristics ->> 'building_class' AS building_class, -- A, B, or C
    core.city,
    core.borough
FROM v_listings AS vl
JOIN listings l ON vl.listing_id = l.id
JOIN properties p ON l.property_id = p.id
JOIN listing_commercial AS lp ON l.id = lp.listing_id
WHERE core.property_type IN ('commercial', 'industrial', 'office', 'retail', 'warehouse', 'income_properties');

### 1.3 Commercial View
Combines for-sale and for-rent commercial listings with extension table data.

In [215]:
%%sql
-- CREATE OR REPLACE VIEW v_commercial AS
SELECT
    l.id                                as listing_id,
    l.listing_id                        as centris_id,
    l.price,
    l.category,
    l.created_at,
    p.property_type,
    p.size_sqft,
    p.municipal_assessment,
    p.floors,
    lc.zoning,
    lc.business_type,
    lc.ceiling_height,
    lc.rent_per_sqft,
    lc.rent_period,
    loc.city,
    loc.borough,
    loc.latitude,
    loc.longitude
FROM listings l
JOIN properties p              ON l.property_id = p.id
JOIN locations loc             ON p.location_id  = loc.id
LEFT JOIN listing_commercial lc ON l.id          = lc.listing_id
WHERE p.property_type IN (
    'commercial', 'industrial', 'office',
    'business', 'income_properties'
);

 * postgresql://postgres:***@localhost/real_estate
1393 rows affected.


,listing_id,centris_id,price,category,created_at,property_type,size_sqft,municipal_assessment,floors,zoning,business_type,ceiling_height,rent_per_sqft,rent_period,city,borough,latitude,longitude
0,14,25716231,2399999.00,for_sale,2026-04-25 19:02:24.073459,commercial,None,1320000.00,NaN,NaN,None,None,None,NaN,Montréal,Rosemont/La Petite-Patrie,45.533008,-73.613975
1,6996,20258843,2690000.00,for_sale,2026-04-26 13:11:37.594894,commercial,None,1708000.00,NaN,NaN,None,None,None,NaN,Montréal,Ahuntsic-Cartierville,45.573685,-73.648739
2,9,20129055,1825000.00,for_sale,2026-04-25 19:02:00.441596,income_properties,3579.00,1193700.00,NaN,NaN,None,None,None,NaN,Montréal,Mercier/Hochelaga-Maisonneuve,45.549242,-73.546628
3,22,26419985,80000.00,for_sale,2026-04-25 19:02:51.457715,business,None,None,NaN,NaN,None,None,None,NaN,Montréal,Le Plateau-Mont-Royal,45.520095,-73.566294
4,27,25226218,3495000.00,for_sale,2026-04-25 19:03:13.039661,commercial,None,1705700.00,NaN,NaN,None,None,None,NaN,Montréal,LaSalle,45.430982,-73.594336
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1388,10662,20842760,1490000.00,for_sale,2026-04-28 04:10:45.274994,income_properties,4059.00,1217100.00,NaN,NaN,None,None,None,NaN,Sainte-Anne-de-Bellevue,NaN,45.402847,-73.947356
1389,1749,26886261,1299000.00,for_sale,2026-04-25 21:10:11.234282,business,None,None,NaN,NaN,None,None,None,NaN,Montréal,Côte-des-Neiges/Notre-Dame-de-Grâce,45.454351,-73.641186
1390,2161,23552221,649000.00,for_sale,2026-04-25 21:43:46.401600,business,None,None,NaN,NaN,None,None,None,NaN,Montréal,Rosemont/La Petite-Patrie,45.576655,-73.568657
1391,10728,18122896,225000.00,for_sale,2026-04-28 05:05:09.786014,business,None,None,NaN,NaN,None,None,None,NaN,Montréal,Ahuntsic-Cartierville,45.545956,-73.673258


## 2. Data Quality Audit

Before any cleaning, we establish a clear picture of data completeness 
across all key features. This guides our imputation and filtering strategy.

### 2.1 Completeness Audit — What Do We Actually Have?

The first question before any modeling is simple: how much of the data is 
actually usable? A column missing 60% of its values cannot be a reliable 
model feature — we need to know this now, not after training.

In [46]:
%%sql
SELECT
    'price'                AS column_name,
    COUNT(price)           AS non_null_count,
    ROUND(COUNT(price) * 100.0 / COUNT(*), 2) AS pct_non_null
FROM v_residential_sale

UNION ALL SELECT 'year_built',       COUNT(year_built),       ROUND(COUNT(year_built) * 100.0       / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'area_sqft',        COUNT(area_sqft),        ROUND(COUNT(area_sqft) * 100.0        / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'bedrooms',         COUNT(bedrooms),         ROUND(COUNT(bedrooms) * 100.0         / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'bathrooms',        COUNT(bathrooms),        ROUND(COUNT(bathrooms) * 100.0        / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'parking_total',    COUNT(parking_total),    ROUND(COUNT(parking_total) * 100.0    / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'total_rooms',      COUNT(total_rooms),      ROUND(COUNT(total_rooms) * 100.0      / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'lot_assessment',   COUNT(lot_assessment),   ROUND(COUNT(lot_assessment) * 100.0   / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'building_assessment', COUNT(building_assessment), ROUND(COUNT(building_assessment) * 100.0 / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'total_taxes',      COUNT(total_taxes),      ROUND(COUNT(total_taxes) * 100.0      / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'school_taxes',     COUNT(school_taxes),     ROUND(COUNT(school_taxes) * 100.0     / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'municipal_taxes',  COUNT(municipal_taxes),  ROUND(COUNT(municipal_taxes) * 100.0  / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'latitude',         COUNT(latitude),         ROUND(COUNT(latitude) * 100.0         / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'longitude',        COUNT(longitude),        ROUND(COUNT(longitude) * 100.0        / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'borough',          COUNT(borough),          ROUND(COUNT(borough) * 100.0          / COUNT(*), 2) FROM v_residential_sale
UNION ALL SELECT 'city',             COUNT(city),             ROUND(COUNT(city) * 100.0             / COUNT(*), 2) FROM v_residential_sale

ORDER BY pct_non_null DESC;

 * postgresql://postgres:***@localhost/real_estate


,column_name,non_null_count,pct_non_null
0,parking_total,9327,100.00
1,longitude,9327,100.00
2,latitude,9327,100.00
3,bathrooms,9308,99.80
4,city,9268,99.37
5,price,9267,99.36
6,total_rooms,9266,99.35
7,bedrooms,9110,97.67
8,year_built,9044,96.97
9,area_sqft,8889,95.30


The residential for-sale view contains **9,322 active listings** across all 
residential property types on Montreal Island. Overall data completeness is 
strong — the average non-null rate across all features sits above 95%, which 
is exceptional for scraped real estate data.

#### Tier 1 — Full coverage (100%)
`parking_total`, `latitude`, `longitude` are fully covered. Coordinates come 
directly from schema.org metadata embedded in every listing page, making them 
the most reliable feature in the dataset. Parking defaults to 0 when absent, 
explaining its perfect coverage.

#### Tier 2 — Near-complete (95–100%)
`bathrooms`, `city`, `price`, `total_rooms`, `bedrooms`, `year_built`, 
`area_sqft` all exceed 95% coverage — sufficient for modeling without concern.

A few structural explanations for the remaining gaps:

- **`year_built`** — Missing values are concentrated in newly constructed or 
  pre-construction condos where the construction year is not yet recorded on 
  the platform. A partial recovery is possible by extracting 4-digit years 
  from the listing description using regex — worth attempting before imputation.

- **`area_sqft`** — The column was designed to capture any available area 
  metric (net area, living area, ground floor area). Remaining gaps represent 
  listings where no area figure was disclosed at all — a structural platform 
  limitation, not a scraping issue.

- **`city`** — Near-complete but not perfect. Given that coordinates are 100% 
  covered, missing city values can be recovered via reverse geocoding as a 
  backfill step. This will be addressed in the cleaning phase.

- **`price`** — Despite 99.36% coverage, price is the single most critical 
  feature in this dataset — it is both our prediction target and a mandatory 
  field on any listing platform. Any missing price warrants investigation 
  before dropping. We will inspect those rows individually to determine whether 
  the gap is a scraping failure or a platform anomaly.

#### Tier 3 — Partial coverage (85–95%)
Financial features — `total_taxes`, `municipal_taxes`, `school_taxes`, 
`building_assessment`, `lot_assessment` — form a coherent lower tier.

Missing values here are not random. Two structural causes explain the pattern:

1. **Recent condos** — Municipal assessments are performed on a fixed cycle. 
   Newly built properties may not yet have an official assessment on record.
2. **Seller disclosure** — Tax and assessment figures are voluntarily disclosed. 
   Some sellers omit them strategically.

For modeling purposes, we will impute missing financial values using the 
**median of similar properties** — same property type, same borough — rather 
than a global median, which would introduce cross-type noise.

- **`borough`** — 85.9% coverage is expected and not a concern. Outside the city of
  Montreal, cities on the Island (Westmount, Kirkland, Beaconsfield, 
  etc.) are independent municipalities with no borough subdivision. Full 
  coordinate coverage means geographic models can use lat/lon directly where 
  borough is absent.

#### Decision

| Feature | Strategy |
|---------|----------|
| `parking_total`, `latitude`, `longitude` | Use as-is |
| `bathrooms`, `bedrooms`, `total_rooms` | Median imputation per property type |
| `year_built` | Regex extraction from description → Affect the current year |
| `area_sqft` | Median imputation per property type or Clustering |
| `city` | Reverse geocoding backfill → drop if unresolvable |
| `price` | Inspect missing rows individually → drop if unrecoverable |
| Financial features | Median imputation per property type and borough |
| `borough` | Use as categorical feature where available, lat/lon otherwise |
| Rows missing >3 critical features | Drop — insufficient information for modeling |

---

*Coverage alone does not guarantee data quality — a column can be 100% 
populated with implausible values. We turn next to the distribution of 
values within each feature to detect outliers and inconsistencies.*

## 2.2 Completeness by Property Type

Global averages can mask property-type-specific patterns. A missing 
`area_sqft` in a condo has a different cause than a missing `area_sqft` 
in a condominium house — and requires a different remedy.


In [47]:
%%sql
SELECT
    property_type,
    COUNT(*)                                        AS total,
    ROUND(COUNT(price) * 100.0       / COUNT(*), 1) AS pct_price,
    ROUND(COUNT(area_sqft) * 100.0   / COUNT(*), 1) AS pct_area,
    ROUND(COUNT(year_built) * 100.0  / COUNT(*), 1) AS pct_year_built,
    ROUND(COUNT(bedrooms) * 100.0    / COUNT(*), 1) AS pct_bedrooms,
    ROUND(COUNT(bathrooms) * 100.0   / COUNT(*), 1) AS pct_bathrooms
FROM v_residential_sale
GROUP BY property_type
ORDER BY total DESC;

 * postgresql://postgres:***@localhost/real_estate


,property_type,total,pct_price,pct_area,pct_year_built,pct_bedrooms,pct_bathrooms
0,condo,5479,99.2,98.1,96.9,97.2,99.8
1,house,1897,99.8,96.6,98.2,99.9,99.9
2,duplex,671,99.4,93.0,96.9,99.4,99.6
3,triplex,564,98.9,92.4,96.5,99.5,99.8
4,quadruplex,248,100.0,90.3,97.2,98.0,99.2
5,condominium_house,219,100.0,36.1,90.4,100.0,100.0
6,quintuplex,149,100.0,91.9,94.6,98.0,100.0
7,loft_studio,97,99.0,99.0,97.9,52.6,99.0
8,mobile_home,3,100.0,33.3,100.0,100.0,100.0



The dataset is heavily skewed toward condos (5,479 listings, 59% of total), 
which reflects the Montreal Island market accurately — the Island is 
predominantly a dense urban condo market, with houses and plexes 
concentrated in specific boroughs like NDG, Rosemont and Ahuntsic.

#### Observations

**Price** is near-complete across all types — the only meaningful gap is 
triplex at 98.9%, which we will investigate individually before dropping.

**Area** shows the most variation across types:

- `condominium_house` at 36.1% is the most concerning gap. This property 
  type sits between a condo and a house — the area descriptor varies 
  between listings (some report living area, some lot area, some both). 
  Our COALESCE chain likely misses cases where neither `net_area` nor 
  `living_area` is present in characteristics. We will inspect a sample 
  of these listings before deciding on imputation.

- `mobile_home` at 33.3% is not a concern — only 3 listings, statistically 
  irrelevant. These will be excluded from modeling.

- Plexes (duplex 93%, triplex 92.4%, quadruplex 90.3%, quintuplex 91.9%) 
  show consistent gaps likely explained by the fact that plex listings 
  often report area per unit rather than total building area — a known 
  ambiguity in the platform's data structure.

**Bedrooms**:

- `loft_studio` at 52.6% is **structural, not a bug**. Studio units have 
  no separate bedroom by definition — the platform leaves the field blank. 
  These will be imputed as **0**, not as the property type median.

- All other types exceed 98% — no concern.

#### Decision — Revised Imputation Strategy by Type

| Property Type | area_sqft strategy | year_built strategy | bedrooms strategy |
|--------------|-------------------|--------------------|--------------------|
| condo | Median per borough | Regex from description → median | Median per borough |
| house | Median per borough | Regex from description → median | Median per borough |
| duplex/triplex/quad/quint | Median per type | Regex from description → median | Median per type |
| condominium_house | Inspect sample → median | Regex from description → median | Median per borough |
| loft_studio | Median per borough | Regex from description → median | **0** |
| mobile_home | **Exclude from modeling** | — | — |

---

*With completeness understood at the property type level, we now turn to 
the second quality dimension — not whether values exist, but whether the 
values that do exist are believable. Outliers in price and area can 
silently distort model performance more than missing values ever would.*

### 2.3 Price Distribution and Outliers
We check for suspicious price values — zeros, extreme outliers, 
and listings priced well outside the realistic Montreal market range.

In [48]:
%%sql
SELECT
    property_type,
    COUNT(*)                                        AS total,
    MIN(price)                                      AS min_price,
    ROUND(AVG(price), 0)                            AS avg_price,
    PERCENTILE_CONT(0.25) WITHIN GROUP (ORDER BY price) AS p25,
    PERCENTILE_CONT(0.50) WITHIN GROUP (ORDER BY price) AS median_price,
    PERCENTILE_CONT(0.75) WITHIN GROUP (ORDER BY price) AS p75,
    PERCENTILE_CONT(0.99) WITHIN GROUP (ORDER BY price) AS p99,
    MAX(price)                                      AS max_price,
    COUNT(*) FILTER (WHERE price < 10000)           AS suspicious_low,
    COUNT(*) FILTER (WHERE price > 10000000)        AS suspicious_high
FROM v_residential_sale
WHERE price IS NOT NULL
GROUP BY property_type
ORDER BY total DESC;

 * postgresql://postgres:***@localhost/real_estate


,property_type,total,min_price,avg_price,p25,median_price,p75,p99,max_price,suspicious_low,suspicious_high
0,condo,5434,48000.00,638728,399000.0,499000.0,679000.0,2966750.0,13849736.00,0,2
1,house,1893,188000.00,1587065,749000.0,1070000.0,1768000.0,9041000.0,29500000.00,0,15
2,duplex,667,299900.00,1002554,776100.0,899000.0,1099000.0,2386000.0,3350000.00,0,0
3,triplex,558,649000.00,1168103,899000.0,1095350.0,1299000.0,2657041.0,3600000.00,0,0
4,quadruplex,248,519000.00,1331920,1036750.0,1200000.0,1399225.0,3765000.0,4800000.00,0,0
5,condominium_house,219,375000.00,938793,596750.0,850000.0,1055350.0,2972421.7,3150000.00,0,0
6,quintuplex,149,849000.00,1667625,1350000.0,1545000.0,1899000.0,3829520.0,4150000.00,0,0
7,loft_studio,96,49900.00,353886,283750.0,319000.0,371250.0,946950.0,1098000.00,0,0
8,mobile_home,3,188000.00,214633,197000.0,206000.0,227950.0,249022.0,249900.00,0,0


#### Observation 

The price distribution is legitimate at both extremes, but for opposite reasons. The high end reflects a real luxury segment concentrated in Westmount, Mont-Royal and Outremont — 17 properties above $10M, all in known affluent locations, all to be kept. The low end is not a residential segment at all — 8 listings below $100K are parking spaces and garage units sold as separate condo fractions, confirmed by area under 10 sqft and explicit descriptions.

Those disguise condos must be removed because a $48,000 "condo" with 6 sqft and no bedrooms would catastrophically distort any price-per-sqft calculation and confuse the model about what drives condo pricing.

***Decision*** — we will apply a minimum price floor of $100,000 for all residential modeling. This removes parking spaces cleanly without touching any legitimate entry-level unit. The 17 luxury properties above $10M are retained — they are real market data and their removal would bias the model against the luxury segment.

### 2.4 Area Distribution and Outliers
Unrealistic area values — zeros, extremely small, or impossibly large — 
will distort price-per-sqft calculations and model performance.

In [18]:
%%sql
SELECT
    property_type,
    COUNT(*)                                            AS total,
    COUNT(area_sqft)                                    AS with_area,
    MIN(area_sqft)                                      AS min_area,
    ROUND(AVG(area_sqft), 0)                            AS avg_area,
    PERCENTILE_CONT(0.50) WITHIN GROUP (ORDER BY area_sqft) AS median_area,
    MAX(area_sqft)                                      AS max_area,
    COUNT(*) FILTER (WHERE area_sqft < 100)             AS suspicious_small,
    COUNT(*) FILTER (WHERE area_sqft > 10000)           AS suspicious_large
FROM v_residential_sale
WHERE area_sqft IS NOT NULL
GROUP BY property_type
ORDER BY total DESC;

 * postgresql://postgres:***@localhost/real_estate


,property_type,total,with_area,min_area,avg_area,median_area,max_area,suspicious_small,suspicious_large
0,condo,4668,4668,4,900,826.5,7367,7,0
1,house,1830,1830,1,6417,3700.0,857823,1,143
2,duplex,624,624,759,2644,2303.0,9725,0,0
3,triplex,521,521,11,2834,2516.0,33077,1,3
4,quadruplex,224,224,28,3419,3126.0,12249,1,2
5,quintuplex,137,137,1,3715,3537.0,11098,1,1
6,loft_studio,89,89,180,483,398.0,1611,0,0
7,condominium_house,79,79,513,2584,2435.0,8020,0,0
8,mobile_home,1,1,560,560,560.0,560,0,0




Area is the second most important price driver after location — a 100 sqft 
difference in a Montreal condo translates to roughly $50,000 in asking price. 
It is also the most contaminated feature in this dataset. Two distinct 
problems emerged from the distribution analysis, each with a different cause 
and a different fix.

---

#### Problem 1 — Building Area (`area_sqft`)

The COALESCE chain in `v_residential_sale` was designed to maximize coverage 
by falling back through multiple area sources:

```sql
COALESCE(net_area, living_area, building_area_at_ground_level, lot_area)
```

This introduced a contamination: for estate and rural properties where no 
building area was recorded, `lot_area` was picked up as the fallback — 
producing values up to **857,823 sqft** that represent land, not structure. 
A 857,823 sqft building does not exist on Montreal Island.

A secondary issue: regex stripping of non-numeric characters from 
characteristics fields produced **single-digit values** for some plexes and 
condos where the source field contained descriptive text rather than a clean 
number — for example "3 floors" being stripped to "3".

**Fix — remove `lot_area` from the COALESCE chain in the view:**

```sql
COALESCE(net_area, living_area, building_area_at_ground_level)
-- lot_area removed — tracked separately as lot_size_sqft
```

After this fix, remaining outliers are handled by thresholding:

---

#### Problem 2 — Lot Size (`lot_size_sqft`)

The lot size distribution is largely clean — realistic medians across all 
plex types and houses reflect known Montreal lot dimensions. Three anomalies 
were identified:

- **3 houses with implausible lot values** — traced to platform-level data 
  entry errors unrecoverable from our end. No alternative source available.

- **15 rows with values consistent with sqm/sqft confusion** — a lot recorded 
  as 500 sqm was likely entered as 500 sqft, producing an underestimate by 
  a factor of ~10.7. At 15 rows, this is too few to justify a bulk conversion 
  rule — the risk of mis-converting legitimate sqft values outweighs the 
  benefit. Treated as NULL and imputed.

---

#### Decision

| Condition | Action |
|-----------|--------|
| `area_sqft < 100` after view fix | NULL → impute per type and borough |
| `area_sqft > 50,000` for houses | NULL → estate segment, exclude from area modeling |
| `lot_size_sqft < 500` | NULL → impute per type and borough |
| `lot_size_sqft > 100,000` | Estate segment — exclude from lot modeling |

The view will be updated to remove `lot_area` from the building area chain 
before the cleaning phase begins. Lot area remains tracked independently 
in `v_plex` and `v_residential_sale` as `lot_size_sqft`.

---

*Area and price outliers are now understood. We turn next to the temporal 
dimension — year built — where the risk is not extreme values but 
systematically wrong ones introduced by the scraper's regex extraction.*

### 2.5 Year Built Distribution
Properties built before 1800 or after the current year are data errors.
We also check the distribution to understand the age profile of the dataset.

In [11]:
%%sql
SELECT
    property_type,
    COUNT(*)                                                AS total,
    MIN(year_built)                                         AS min_year,
    ROUND(AVG(year_built), 0)                               AS avg_year,
    PERCENTILE_CONT(0.50) WITHIN GROUP (ORDER BY year_built) AS median_year,
    MAX(year_built)                                         AS max_year,
    COUNT(*) FILTER (WHERE year_built < 1800)               AS suspicious_old,
    COUNT(*) FILTER (WHERE year_built > 2026)               AS suspicious_future
FROM v_residential_sale
WHERE year_built IS NOT NULL
GROUP BY property_type
ORDER BY total DESC;

 * postgresql://postgres:***@localhost/real_estate


,property_type,total,min_year,avg_year,median_year,max_year,suspicious_old,suspicious_future
0,condo,5310,1800,1997,2012.0,2026,0,0
1,house,1862,1770,1965,1965.0,2026,2,0
2,duplex,650,1770,1941,1949.0,2026,1,0
3,triplex,544,1870,1941,1947.5,2026,0,0
4,quadruplex,241,1800,1949,1956.0,2021,0,0
5,condominium_house,198,1840,1993,2002.0,2025,0,0
6,quintuplex,141,1870,1939,1931.0,2025,0,0
7,loft_studio,95,1900,2002,2018.0,2025,0,0
8,mobile_home,3,1973,1980,1975.0,1992,0,0


Year built serves two roles in the model: as a direct age feature 
(older buildings command different prices) and as a proxy for renovation 
cycle, structural condition, and building standards. A 1940s duplex and 
a 2020 condo are fundamentally different assets — the model needs to 
distinguish them reliably.

Overall the distribution is coherent and reflects Montreal's urban history 
well. The city's residential stock was built in three distinct waves:

- **Pre-1950** — the plex era. Duplexes (median 1949), triplexes (1947), 
  quintuplexes (1931) and houses (1965) reflect the dense working-class 
  neighbourhoods built along streetcar lines in Rosemont, Plateau, 
  Hochelaga and Villeray. These medians are historically accurate.

- **1950–2000** — the suburban expansion. Houses and quadruplexes 
  fill this period as Montreal expanded westward into NDG, Côte-des-Neiges 
  and the West Island.

- **Post-2000** — the condo boom. Condos (median 2012) and loft studios 
  (median 2018) dominate recent construction, consistent with the 
  high-rise development wave along the downtown core and Griffintown.

#### Suspicious values

**Pre-1800 entries** — 3 houses and 1 duplex were built before 1800 
according to the platform. Montreal was founded in 1642 and its oldest 
surviving residential buildings date to the late 1700s. While not 
technically impossible, these values are almost certainly data entry 
errors on the platform — a seller entering "1700" instead of "1970" 
is a common transcription mistake. These 3 rows represent less than 
0.03% of the dataset and will be set to NULL before imputation.

**Max year = 2026** — Present across condo, house, duplex and triplex. 
These are pre-construction or under-construction listings where the 
platform records the expected completion year. This is valid data, 
not an error — 2026 means "delivering this year". No action needed.

**Condo minimum = 1800** — The 1800 floor for condos warrants a closer 
look. A condo built in 1800 predates the legal concept of condominium 
ownership in Quebec. This is likely a heritage building converted to 
condos where the original construction date was entered rather than 
the conversion date. A small number of such cases are expected in 
Montreal's Old Port and Vieux-Montréal neighbourhoods.

#### Decision

| Condition | Action |
|-----------|--------|
| `year_built < 1850` for duplex, triplex, quintuplex | Inspect individually — likely transcription error → NULL |
| `year_built < 1800` for house | NULL → impute per borough |
| `year_built < 1800` for condo | Retain if Old Port borough, else NULL |
| `year_built = 2026` | Retain — valid pre-construction data |
| Remaining NULLs after inspection | Median imputation per property type and borough |

---

*Year built is now understood. With completeness and individual feature 
distributions audited, we shift from examining features in isolation to 
examining how they relate to each other — and most importantly, how they 
relate to price.*

### 2.6 Geographic Coverage
We verify that coordinates and borough data are available for 
enough listings to support geographic analysis and location-based features.

In [12]:
%%sql
SELECT
    city,
    COUNT(*)                                            AS total,
    COUNT(borough)                                      AS with_borough,
    COUNT(latitude)                                     AS with_coords,
    COUNT(*) - COUNT(latitude)                          AS missing_coords,
    ROUND(COUNT(latitude) * 100.0 / COUNT(*), 1)        AS pct_geocoded
FROM v_residential_sale
GROUP BY city
ORDER BY total DESC;

 * postgresql://postgres:***@localhost/real_estate


,city,total,with_borough,with_coords,missing_coords,pct_geocoded
0,Montréal,7978,7978,7978,0,100.0
1,Westmount,200,0,200,0,100.0
2,Pointe-Claire,192,0,192,0,100.0
3,Dollard-des-Ormeaux,183,0,183,0,100.0
4,Mont-Royal,165,0,165,0,100.0
5,Côte-Saint-Luc,140,0,140,0,100.0
6,Beaconsfield,109,0,109,0,100.0
7,Kirkland,84,0,84,0,100.0
8,Dorval,61,0,61,0,100.0
9,NaN,59,34,59,0,100.0


Location is the dominant price driver in real estate — "location, location, 
location" is not a cliché, it is an empirical regularity confirmed in every 
hedonic pricing study. A dataset with poor geographic coverage cannot support 
meaningful spatial analysis. Fortunately, this is where our pipeline performs 
best.

#### Coordinate coverage — perfect

Every single listing in the dataset is geocoded — **100% coordinate coverage 
across all 15 cities**. This is a direct result of the scraping architecture: 
coordinates are extracted from schema.org metadata embedded in every listing 
page, making them the most reliably collected feature in the entire pipeline.

#### Borough coverage — by design, not by accident

The borough column tells a story about Montreal Island's political geography 
rather than a data quality issue.

**Montréal** (7,978 listings, 86% of dataset) has full borough coverage — 
every listing is assigned to one of its 19 administrative boroughs (Plateau, 
Rosemont, Outremont, Ville-Marie, etc.). Borough is a meaningful and 
granular geographic feature for Montréal listings.

**All other cities have zero borough coverage** — and this is correct. 
Westmount, Pointe-Claire, Dollard-des-Ormeaux, Mont-Royal and the other 
West Island municipalities are independent cities, not boroughs of Montréal. 
They have no internal borough subdivision. Filling borough as NULL for these 
listings is the right behavior.

This means borough is only usable as a feature for the Montréal subset. 
For the full island-wide model, `city` is the appropriate geographic 
categorical — and it covers 99.4% of listings.

#### The blank city row

One row stands out — 59 listings with no city value, 34 of which have a 
borough. This is the parsing bug flagged in section 2.1. These are Montréal 
listings where the address parser extracted the borough but failed to assign 
the city. Since the borough is present and coordinates are available, these 
59 listings can be recovered:

```sql
UPDATE locations
SET city = 'Montréal'
WHERE city IS NULL
AND borough IS NOT NULL;
```

This will be applied in the cleaning phase.

#### Decision

| Condition | Action |
|-----------|--------|
| `city IS NULL` and `borough IS NOT NULL` | Backfill `city = 'Montréal'` |
| `city IS NULL` and `borough IS NULL` | Use coordinates for spatial features only |
| `borough IS NULL` outside Montréal | Expected — use `city` as geographic feature |
| Full island model | Use `city` as primary geographic categorical |
| Montréal-only model | Use `borough` as primary geographic categorical |

---

*Geographic coverage is complete and the city parsing gap is recoverable. 
With all individual features audited, we now have enough information to 
define our clean modeling dataset — the rows and columns that will actually 
enter the price prediction model.*

### 2.7 Condo Extension — Completeness

In [15]:
%%sql
SELECT
    'platform_id'                AS column_name,
    COUNT(platform_id)           AS non_null_count,
    ROUND(COUNT(platform_id) * 100.0 / COUNT(*), 2) AS pct_non_null
FROM v_condo

UNION ALL SELECT 'condo_fees_yearly',       COUNT(condo_fees_yearly),       ROUND(COUNT(condo_fees_yearly) * 100.0       / COUNT(*), 2) FROM v_condo
UNION ALL SELECT 'unit_number',        COUNT(unit_number),        ROUND(COUNT(unit_number) * 100.0        / COUNT(*), 2) FROM v_condo
UNION ALL SELECT 'condo_type',         COUNT(condo_type),         ROUND(COUNT(condo_type) * 100.0         / COUNT(*), 2) FROM v_condo
UNION ALL SELECT 'locker',        COUNT(locker),        ROUND(COUNT(locker) * 100.0        / COUNT(*), 2) FROM v_condo

ORDER BY pct_non_null DESC;

 * postgresql://postgres:***@localhost/real_estate


,column_name,non_null_count,pct_non_null
0,platform_id,5698,100.00
1,condo_type,5698,100.00
2,locker,5698,100.00
3,condo_fees_yearly,5569,97.74
4,unit_number,4756,83.47


In [13]:
%%sql
SELECT
    COUNT(*)                                            AS total_condos,
    COUNT(condo_fees_yearly)                            AS with_condo_fees,
    COUNT(unit_number)                                  AS with_unit_number,
    COUNT(*) FILTER (WHERE locker = true)               AS with_locker,
    COUNT(condo_type)                                   AS with_condo_type,
    COUNT(*) - COUNT(condo_fees_yearly)                 AS missing_fees,
    COUNT(*) - COUNT(unit_number)                       AS missing_unit
FROM v_condo;

 * postgresql://postgres:***@localhost/real_estate


,total_condos,with_condo_fees,with_unit_number,with_locker,with_condo_type,missing_fees,missing_unit
0,5698,5569,4756,736,5698,129,942



The condo extension table adds three features critical to condo valuation:
monthly fees, unit floor, and storage. These are condo-specific costs and 
amenities that have no equivalent in house or plex pricing — a condo with 
$1,200/month fees is a fundamentally different investment than an identical 
unit with $300/month fees, regardless of asking price.

#### Observations

**`condo_fees_yearly`** at 97.74% coverage is strong. The 2.26% gap 
(approximately 129 listings) likely represents newly launched projects 
where the condo syndicate has not yet set its fee schedule, or older 
listings where the seller omitted this figure. Given the strong coverage, 
median imputation per `condo_type` and `borough` is appropriate for the 
missing cases.

**`locker`** and **`condo_type`** at 100% — both default to a value when 
absent (`locker = FALSE`, `condo_type` extracted from characteristics). 
The defaults are sensible and no action is needed.

**`unit_number`** at 83.47% is the most notable gap. Three causes explain 
this:

- **Condominium houses** — townhouse-style condos where the civic number 
  serves as the unit identifier. No separate unit number exists on the 
  platform.
- **Ground floor commercial condos** — some mixed-use condo listings do 
  not carry a unit number.
- **Parsing gaps** — our address parser may have missed unit numbers 
  embedded in non-standard formats such as "1055B Rue Saint-Mathieu" 
  where the letter suffix serves as the unit identifier.

`unit_number` is not a model feature — it serves as a deduplication key 
in the location table. Its missingness does not affect modeling but does 
affect our ability to distinguish two units in the same building. Where 
missing, coordinates provide sufficient geographic granularity.

#### Decision

| Condition | Action |
|-----------|--------|
| `condo_fees_yearly IS NULL` | Median imputation per `condo_type` and `borough` |
| `locker = FALSE` | Retain — valid default |
| `unit_number IS NULL` | No action for modeling — use coordinates |
| `condo_type IS NULL` | Not possible — 100% covered |

---

*The condo extension is clean and well-covered. We turn next to the plex 
extension where the key variable — rental income — carries the entire 
weight of investment yield analysis for multi-unit properties.*

### 2.8 Plex Extension — Completeness and Income Coverage
Rental income is the key metric for plex investment analysis.
We check coverage and flag implausible values.

In [22]:
%%sql
SELECT
    property_type,
    COUNT(*)                                            AS total,
    COUNT(units)                                        AS with_units,
    COUNT(rental_income)                                AS with_rental_income,
    COUNT(living_area)                                  AS with_living_area,
    ROUND(AVG(rental_income), 0)                        AS avg_rental_income,
    PERCENTILE_CONT(0.50) WITHIN GROUP 
        (ORDER BY rental_income)                        AS median_rental_income,
    COUNT(*) FILTER (WHERE rental_income < 1000)        AS suspicious_low_income,
    COUNT(*) FILTER (WHERE rental_income > 500000)      AS suspicious_high_income
FROM v_plex
GROUP BY property_type
ORDER BY total DESC;

 * postgresql://postgres:***@localhost/real_estate


,property_type,total,with_units,with_rental_income,with_living_area,avg_rental_income,median_rental_income,suspicious_low_income,suspicious_high_income
0,duplex,671,671,671,210,49006,46200.0,4,0
1,triplex,564,564,564,172,64166,57828.0,0,0
2,quadruplex,248,248,247,72,71027,66060.0,0,0
3,quintuplex,149,149,149,50,94063,85992.0,0,0



For plex properties, the asking price is only half the story. A triplex 
listed at $900,000 generating $72,000 in annual rental income is a 
fundamentally different investment than one generating $42,000 — the 
gross yield gap between these two scenarios is the difference between 
a viable investment and a money-losing one. Rental income is therefore 
the single most important feature in plex analysis, more so than any 
physical characteristic.

#### Unit count — perfect coverage

All four plex types show 100% unit count coverage — duplex always has 2, 
triplex 3, and so on. This is expected since unit count is definitional 
to the property type itself and is always disclosed on the platform.

#### Rental income — strong but not perfect

Coverage is near-complete across all types — 671/671 duplexes, 564/564 
triplexes, 248/248 quadruplexes and 149/149 quintuplexes carry a rental 
income figure. The single missing quadruplex value warrants individual 
inspection before imputation.

The income distribution tells a coherent story about Montreal's plex market:

| Type | Median Annual Income | Per Unit |
|------|---------------------|----------|
| Duplex | $46,200 | $23,100 |
| Triplex | $57,828 | $19,276 |
| Quadruplex | $66,060 | $16,515 |
| Quintuplex | $85,992 | $17,198 |

Per-unit income decreases slightly as unit count grows — consistent with 
Montreal market dynamics where larger plexes tend to be located in 
working-class neighbourhoods with lower rents per unit, while duplexes 
are more evenly distributed across price segments including premium 
boroughs like Outremont and Plateau.

#### Suspicious low income — 4 duplexes

Four duplexes flagged with implausibly low rental income. Three scenarios 
explain low duplex income:

1. **Owner-occupied unit** — the owner lives in one unit and rents only 
   the other, with the platform recording income from one unit only
2. **Below-market lease** — a long-term tenant on a pre-rent-control 
   lease paying well below market rate
3. **Data entry error** — annual income entered as monthly, or a digit 
   dropped

These 4 rows represent 0.6% of duplexes. Given the ambiguity between 
legitimate below-market situations and errors, we will set these to NULL 
and impute rather than drop — preserving the rows for price modeling 
while excluding the income figure from yield calculations.

#### Living area — the real gap

`with_living_area` reveals the most significant gap in the plex extension:
210/671 duplexes (31%), 172/564 triplexes (30%), 72/248 quadruplexes (29%) 
and 50/149 quintuplexes (34%). Roughly 70% of all plexes are missing 
living area.

This is a structural platform issue — plex listings on Centris typically 
report area per unit rather than total building area, and our scraper 
captures total building area. When no total is disclosed, the field is 
empty. This cannot be recovered without revisiting individual listings.

For plex modeling, we will use `units × median_unit_area_per_type` as a 
synthetic area feature where living area is missing — a reasonable 
approximation given the near-uniform unit counts per type.

#### Decision

| Condition | Action |
|-----------|--------|
| `rental_income < 10,000` for duplex | NULL → impute per borough |
| Single missing quadruplex income | Inspect individually → impute |
| `living_area IS NULL` | Synthetic: `units × median unit area per type` |
| All rental income values | Verify annual vs monthly — check against price for yield sanity |

---

*The plex extension is reliable where it matters most — rental income 
is near-complete and the distribution is internally consistent with 
known Montreal market dynamics. With all extension tables audited, 
we now have a complete picture of data quality across the full 
residential dataset. The next step is translating every decision 
made in this audit into concrete SQL cleaning operations.*

## 2.9 Audit Summary — What We Are Working With

A data audit is only useful if it produces decisions. This section 
consolidates every finding from sections 2.1 through 2.8 into a single 
clear picture of what the dataset looks like before cleaning, and what 
it will look like after.

### The dataset before cleaning

We started with **9,322 active residential for-sale listings** across 
Montreal Island, scraped across multiple runs to approach saturation on 
the Centris platform. On the surface, this is an unusually clean scraped 
dataset — 95%+ average coverage, 100% geocoding, and a coherent 
distribution that reflects known Montreal market dynamics.

Beneath the surface, four categories of issues were identified:

**Structural gaps** — missingness that is not a bug but a feature of 
the market itself. Borough is absent outside Montréal because other 
Island cities have no borough subdivision. Loft studios have no bedroom 
count because studios have no separate bedroom. Year built is missing 
for pre-construction condos because the building does not yet exist. 
These gaps carry information — they should be handled with domain 
knowledge, not mechanical imputation.

**Platform limitations** — missingness that reflects what sellers choose 
to disclose. Tax figures, assessment values and condo fees are 
voluntarily provided. A seller omitting their municipal taxes is making 
a choice. These fields average 90–93% coverage — strong enough to use, 
but the missing cases are not random and median imputation must be 
applied at the borough and property type level to avoid introducing 
cross-market noise.

**Scraping artifacts** — a small number of issues traceable to our own 
pipeline. The COALESCE chain in `v_residential_sale` contaminated 
`area_sqft` with lot area values for estate properties. The address 
parser failed to assign city for 59 Montréal listings where borough 
was successfully extracted. Single-digit area values appeared where 
regex stripping encountered non-numeric text. These are recoverable 
and will be fixed at the view level before any modeling begins.

**True outliers** — values that exist in the data but cannot be real. 
Three houses built before 1800, four duplexes with implausibly low 
rental income, fifteen listings with likely sqm/sqft confusion in lot 
size. These represent less than 0.1% of the dataset combined — too few 
to affect model performance if handled correctly, too dangerous to leave 
in place if not.

### The dataset after cleaning

After applying every decision documented in this audit, the expected 
modeling dataset is:

| Segment | Raw listings | Expected after cleaning | Loss |
|---------|-------------|------------------------|------|
| Full residential | 9,322 | ~9,100 | ~2% |
| Condo model | 5,479 | ~5,350 | ~2% |
| House model | 1,897 | ~1,850 | ~2.5% |
| Plex model | 1,632 | ~1,580 | ~3% |
| Plex yield analysis | 1,632 | ~1,570 | ~3.5% |

A 2–3% reduction from cleaning is the expected cost of quality. Losing 
more than 10% would signal either over-aggressive filtering or a 
fundamental data collection problem — neither applies here.

### Features entering the model

After cleaning, the confirmed feature set per model is:

**Price prediction — all residential:**
`price` (target), `property_type`, `area_sqft`, `bedrooms`, `bathrooms`, 
`parking_total`, `garage`, `pool`, `basement`, `year_built`, 
`municipal_assessment`, `total_taxes`, `city`, `borough` (Montréal only), 
`latitude`, `longitude`

**Condo model additions:**
`condo_fees_yearly`, `locker`, `condo_type`, `unit_number`

**Plex model additions:**
`units`, `rental_income`, `living_area` (synthetic where missing)

**Features excluded from modeling:**
`move_in_date` — administrative, not a price driver  
`total_rooms` — collinear with `bedrooms` and `bathrooms`  
`lot_size_sqft` — estate segment only, excluded from main model  
`description` — unstructured text, reserved for NLP extension  
`mobile_home` — 3 listings, statistically irrelevant  

### What this audit tells us about the pipeline

Beyond the immediate modeling decisions, this audit surfaces two 
improvements for the scraping pipeline before the next city is added:

1. **Remove `lot_area` from the building area COALESCE chain** — this 
   fix belongs in `v_residential_sale` and prevents estate lot values 
   from contaminating building area going forward.

2. **Fix the city parser for borough-only addresses** — 59 listings 
   recovered manually in this audit should be recovered automatically 
   by the scraper on future runs.

Both fixes are low-effort and high-impact. They will be applied to 
the view definition and the address parser before Laval scraping begins.

---

*The audit is complete. Every feature is understood, every gap has a 
strategy, and every decision is documented. We move now to the cleaning 
phase — translating these decisions into SQL operations that produce 
the final modeling dataset.*